feat(connectors): migrate an old connector to the new format (#1635)
bonitasoft/bonita-doc: modules/ROOT/pages/connector-migration.adoc

> = Migrate an old connector created with Bonita Studio to a dedicated connector project

Reviewer: ```suggestion
= Migrate a connector created with Bonita Studio to a dedicated project
```

> :description: Migrate an old connector created with Bonita Studio to a maven archetype style Connector project

Reviewer: ```suggestion
:description: Migrate a connector created with Bonita Studio to a dedicated project using the Maven archetype
```

> Since Bonita 2021.2, Connectors cannot be created from the Studio anymore. Developers have to create dedicated connector projects outside the Studio, in their favorite IDE, to develop build and release their connectors. Then, connectors can be added to Bonita projects as xref:software-extensibility.adoc[extensions]. This allow connectors to have their own lifecycle, developers can set up proper continuous integrations for connector projects and test them correctly. +

Reviewer: ```suggestion
Since Bonita 2021.2, Connectors cannot be created from the Studio anymore. Developers have to create dedicated connector projects outside the Studio, in their favorite IDE, to develop, build and release their connectors. Then, connectors can be added to Bonita projects as xref:software-extensibility.adoc[extensions]. This allow connectors to have their own project lifecycle and let the developers uses the latest state of the art tooling to manage the continous integration and code quality. +
```

> This page describes the way to migrate an old custom connector created with Bonita Studio into a dedicated connector project created with the provided maven archetype.

Reviewer: ```suggestion
This page describes how to migrate a custom connector created with Bonita Studio into a dedicated connector project using the provided maven archetype.
```

[NOTE]
====

> All the content of this page can be applied to migrate an old actor filter created with Bonita Studio to a dedicated actor filter project. The actor filter archetype has to be used instead of the connector archetype.

Reviewer: ```suggestion
All the content of this page can be applied to migrate actor filters created with Bonita Studio to a dedicated actor filter project. Use the xref:actor-filter-archetype.adoc[actor filter archetype] instead.
```

For example, if a connector depends on the Jar `commons-lang3-3.12.0.jar`, then the corresponding Maven dependency is the dependency `Apache Commons Lang` in version `3.12.0`. So, the following Maven dependency has to be declared in the `pom.xml`: 

[source, xml]
----
<dependency>
    <groupId>org.apache.commons</groupId>
    <artifactId>commons-lang3</artifactId>
    <version>3.12.0</version>
</dependency>
----

Finding the corresponding maven dependency of a Jar file can be hard, even impossible if the Jar has never been deployed on any Maven repository.

The first thing to do is to simply copy past the Jar name in a web search engine, and hopefully the corresponding maven dependency will show up. You can also try to look for it on famous public maven repositories, like https://mvnrepository.com/[Maven central].

In case it doesn't work, you can try to explore the Jar file (using an archive manager) to see which class it contains. +
For example, if you explore the jar `commons-lang3-3.12.0.jar`, you will find a class named `StringUtils`. If you type _StringUtils maven_ in a web search engine you have a good chance to find the corresponding maven dependency.

> If the corresponding Maven dependency cannot be found publicly, it can mean that it is a private dependency of your organization. In this case, ask your IT team a way to retrieve this dependency through Maven, and to consider to deploy it on a maven repository (private or public) if it's not the already case.

Reviewer: Not sure if it is relevant here

Author: I disagree, I'm pretty sure it can often happen.

If you can't find a way to retrieve the corresponding maven dependency of a Jar file, you should consider to drop it for an other one available publicly. Try to find out in your source code the purpose of this dependency, and look for an equivalent available. In this case, you will have ot update the implementation code of your connector.

[IMPORTANT]
====
Always consider to drop a dependency that is not available publicly and maintained, unless it's your dependency (so you can maintain it if needed). +

> Beeing enable to update a dependency means that your application can stop to work from one day to another, following an OS update, a Java update, and you won't have any solution to fix it easily.

Reviewer: ```suggestion
Not beeing able to update a dependency means that your application is exposed to maintainability and security issues (eg: OS update, Java update...).
```

====

== 3 - Retrieve the definition content

> The next step is to retrieve the definition content from the old project an to apply it to the new project. To do so, a part of the old _.def_ file will be copied into the new _.def_ file. 

Reviewer: ```suggestion
The next step is to retrieve the definition content from the old project and to copy it to the new project. To do so, a part of the old _.def_ file will be copied into the new _.def_ file. 
```

> Start to open the old _.def_ file (usually _definitionId-version.def_) in a text editor. +

Reviewer: ```suggestion
Open the old _.def_ file (usually _definitionId-version.def_) in a text editor. +
```

    <icon>connector.png</icon> <!-- The icon used in the Studio for this definition -->
    <category icon="connector.png" id="Custom"/> <!-- The category of this definition, used in the Studio (e.g: http, script ...) -->

    <!-- Connector inputs -->
  <input defaultValue="defaultValue" mandatory="true" name="Input1" type="java.lang.String"/> <1>
  <input mandatory="true" name="Input2" type="java.lang.String"/>

    <!-- Connector outputs -->
    <output name="Output1" type="java.lang.String"/> <2>

    <!--
       Pages and widgets to use the connector in the Bonita Studio.
       - Each widget must be bound to an input
       - Page titles must be defined in the properties files
       - Widget labels must be defined in the properties files
       - Page and widget descriptions can be defined in the properties files (optional)
    -->
    <page id="input-page"> <3>
        <widget xsi:type="definition:Text" id="name_widget" inputName="name"/>
    </page>
</definition:ConnectorDefinition>
----
<1> The inputs of the connector
<2> The outputs of the connector
<3> The pages and widgets of the connector, used by the Studio to configure a call to the connector.

> The icons and the category can also be transfered into the new definition, if so then put the corresponding icon files in _src/main/resources_.

Reviewer: ```suggestion
The icons and the localization files (_<definition-id>*.properties_) should also be copied into the new project resources folder:  _src/main/resources_.
```

[NOTE]
====
All the content of this page can be applied to migrate actor filters created with Bonita Studio to a dedicated actor filter project. Use the xref:actor-filter-archetype.adoc[actor filter archetype] instead.
====

== 1 - Create a new connector project using the maven archetype

First of all, a new connector project has to be created using the provided maven archetype. +
xref:connector-archetype#_generate_the_project_using_the_maven_archetype[This piece of documentation] explains how to create a connector project using the maven archetype.

Some parameters have to be passed to the maven archetype to create the project (groupid, artifactId, version, package, language...). Make sure to use *Java* as the language, so the existing implementation code will be reusable. It is also recommended to keep the package of the current connector implementation, else you might need to edit some part of your implementation source code.

[TIP]
====
To ease the migration, it is recommended to use the same definition id and definition version of the current connector. +

> To do so, the maven properties `<connector-definition-id>` and `<connector-definition-version>` have to be edited in the generated pom.xml of the project with the correct values. By default, it takes the project artifact id as definition id and the project version as definition version.

Reviewer: ```suggestion
To do so, the maven properties `connector-definition-id` and `connector-definition-version` have to be edited in the generated pom.xml of the project with the correct values. By default, it takes the project artifact id as definition id and the project version as definition version.
```

Now, connector projects are Maven projects, dependencies are handled by Maven. It means that all the connector dependencies have to be declared in the file `pom.xml` (at the root of the project) as maven dependencies. https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html[Learn more about the Maven dependency mechanism].

For each declared dependency in your implementation or in your definition, *the corresponding Maven dependency has to be found and declared in the pom.xml*. +
For example, if a connector depends on the Jar `commons-lang3-3.12.0.jar`, then the corresponding Maven dependency is the dependency `Apache Commons Lang` in version `3.12.0`. So, the following Maven dependency has to be declared in the `pom.xml`: 

[source, xml]
----
<dependency>
    <groupId>org.apache.commons</groupId>
    <artifactId>commons-lang3</artifactId>
    <version>3.12.0</version>
</dependency>
----

Finding the corresponding maven dependency of a Jar file can be hard, even impossible if the Jar has never been deployed on any Maven repository.

> The first thing to do is to simply copy past the Jar name in a web search engine, and hopefully the corresponding maven dependency will show up. You can also try to look for it on famous public maven repositories, like https://mvnrepository.com/[Maven central].

Reviewer: ```suggestion
The first thing to do is to copy/paste the jar name in a web search engine, and hopefully the corresponding maven dependency will show up. You can also try to look for it on public maven repositories, like https://mvnrepository.com/[Maven central].
```

		return (java.lang.String) getInputParameter(INPUT1_INPUT_PARAMETER);
	}

	protected final void setOutput1(java.lang.String output1) {
		setOutputParameter(OUTPUT1_OUTPUT_PARAMETER, output1);
	}

	@Override
	public void validateInputParameters() throws ConnectorValidationException {
		try {
			getInput1();
		} catch (ClassCastException cce) {
			throw new ConnectorValidationException("input1 type is invalid");
		}

	}

}
----

.Old connector class, to execute the business logic
[source, Java]
----
package org.mycompany.connector;

import org.bonitasoft.engine.connector.ConnectorException;

public class MyConnectorDefImpl extends AbstractMyConnectorDefImpl {

	@Override
	protected void executeBusinessLogic() throws ConnectorException {
		String input1 = getInput1();
		// [...]
		setOutput1("output");
	 }

	@Override
	public void connect() throws ConnectorException {
	}

	@Override
	public void disconnect() throws ConnectorException{
	}

}
----

In the new connector projects, generated by the Maven archetype, there is only one class by default which extends `org.bonitasoft.engine.connector.AbstractConnector`, located in `src/main/<your package>`. 

> Now it's up to you, you can either delete the provided class and copy past your old classes, or just pick the content from the old classes and past it in the new one. +

Reviewer: ```suggestion
Now it's up to you, you can either delete the provided class and copy/paste your old classes, or just pick the content from the old classes and paste it in the new one. +
```

        }
    }

    @Override
    protected void executeBusinessLogic() throws ConnectorException {
        String input1 = getInputParameter(INPUT1_INPUT_PARAMETER);
        LOGGER.info(String.format("Default input: %s", input1));
        // [...]
        setOutputParameter(DEFAULT_OUTPUT, "output");
    }

    @Override
    public void connect() throws ConnectorException{}

    @Override
    public void disconnect() throws ConnectorException{}
}
----

[CAUTION]
====
The main class of your connector is referenced in the _.impl_ file. +
Make sure that at the end of the operation, the _.impl_ file points to the main implementation class
====

== 6 - Update the unit tests

With the new connector project format comes the possibility to write unit tests for your connectors. +
Some default tests are generated by the Maven archetype, in `src/test/java/<your package>`. Those tests perform some validations on the default inputs generated. You'll have to update those tests to make them pass, according to the content of your connector project. +
For example, it is a good practice to test that the connector execution fails properly if a mandatory input is missing.

> It is also possible to just delete the test file, so the project will build.

Reviewer: ```suggestion
It is also possible to just delete the test file to make the project build.
```
